pkg/syncutil/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,92 @@
+package cachemanager
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/controller/config/process"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/logging"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/metrics"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/syncutil"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("data-replication").WithValues("metaKind", "CacheManagerTracker")
+
+type CacheManager struct {
+	opa              syncutil.OpaDataClient
+	syncMetricsCache *syncutil.MetricsCache
+	tracker          *readiness.Tracker
+	processExcluder  *process.Excluder
+
+	// todo acpana -- integrate gvkaggregator
+}
+
+func NewCacheManager(opa syncutil.OpaDataClient, syncMetricsCache *syncutil.MetricsCache, tracker *readiness.Tracker, processExcluder *process.Excluder) *CacheManager {
+	return &CacheManager{
+		opa:              opa,
+		syncMetricsCache: syncMetricsCache,
+		tracker:          tracker,
+		processExcluder:  processExcluder,
+	}
+}
+
+func (c *CacheManager) AddObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	isNamespaceExcluded, err := c.processExcluder.IsNamespaceExcluded(process.Sync, instance)
+	if err != nil {
+		return fmt.Errorf("error while excluding namespaces: %w", err)
+	}
+
+	// bail because it means we should not be
+	// syncing this gvk
+	if isNamespaceExcluded {
+		c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+		return c.RemoveObject(ctx, instance)
```

Comment:
While this line isn't harmful, testing it below is adding overhead without gain (at least that I see).

Explanation:

Any time that the processExcluder changes, the cache is wiped.  Even in the future (when we can look into the cache), we'll be wiping many objects at once (whole GVKs or perhaps whole namespaces if we can do that).

Thus, any time objects need to be removed from the cache, we'll be doing that removal as a separate operation on all the objects that should be removed.  An item-by-item approach here doesn't really make sense, as we can't reliably do our cleanup this way (even if we have this, we can't know that everything that needs to be removed has been removed, so we'll need to go cleanup on our own anyways).

That's not a big deal necessarily, but adding functionality that has a not necessarily helpful side effect that we then test seems to be adding complexity and cost without benefit.  (not to mention [Hyrum's Law](https://www.hyrumslaw.com/))